File: patterns/_tests/BehavioralPatterns/TemplateMethod/typescript/unit.test.ts
// ============================================================
//  Unit Tests — Template Method: Data Pipeline (TypeScript)
//
//  Run: npx mocha --require ts-node/register unit.test.ts
// ============================================================

import { describe, it } from 'mocha';
import assert from 'assert';
import { captureOutput } from '../../../_helpers/typescript/helpers';
import {
  DataPipeline, CsvPipeline, JsonPipeline, XmlPipeline, type Row,
} from '../../../../BehavioralPatterns/TemplateMethod/typescript/typescript';

// ── Helper: minimal spy pipeline ─────────────────────────────
function makePipeline(records: Row[], name = 'TestPipe'): DataPipeline & { errs: string[] } {
  class SpyPipe extends DataPipeline {
    get name() { return name; }
    protected extract()   { this.records = records; }
    protected transform() { /* no-op */ }
    get errs() { return this.errors; }
    // expose protected methods for direct testing
    runValidate() { (this as any).extract(); (this as any).validate(); }
    runLoad()     { (this as any).load(); }
    runReport()   { (this as any).report(); }
  }
  return new SpyPipe() as any;
}

// ── validate() ───────────────────────────────────────────────
describe('DataPipeline.validate()', () => {
  it('passes records that have both id and name', () => {
    const p = makePipeline([{ id: '1', name: 'Alice' }]);
    captureOutput(() => (p as any).runValidate());
    assert.strictEqual(p.errs.length, 0);
  });

  it('catches a missing name', () => {
    const p = makePipeline([{ id: '1', name: null }]);
    captureOutput(() => (p as any).runValidate());
    assert.ok(p.errs.some((e: string) => e.includes('name')));
  });

  it('catches a missing id', () => {
    const p = makePipeline([{ id: null, name: 'Alice' }]);
    captureOutput(() => (p as any).runValidate());
    assert.ok(p.errs.some((e: string) => e.includes("'id'")));
  });

  it('output line contains "records checked"', () => {
    const p = makePipeline([{ id: '1', name: 'Alice' }]);
    const lines = captureOutput(() => (p as any).runValidate());
    assert.ok(lines.some(l => l.includes('records checked')));
  });
});

// ── report() hook ────────────────────────────────────────────
describe('DataPipeline.report() hook', () => {
  it('prints "No issues." when no errors', () => {
    const p = makePipeline([]);
    const out = captureOutput(() => (p as any).runReport()).join('\n');
    assert.ok(out.includes('No issues.'));
  });

  it('prints warning symbol when errors exist', () => {
    const p = makePipeline([]);
    (p as any).errors = ["row 0: missing required field 'name'"];
    const out = captureOutput(() => (p as any).runReport()).join('\n');
    assert.ok(out.includes('⚠'));
  });

  it('can be overridden in a subclass', () => {
    class CustomCsv extends CsvPipeline {
      protected override report() { console.log('CUSTOM'); }
    }
    const out = captureOutput(() => new CustomCsv().run()).join('\n');
    assert.ok(out.includes('CUSTOM'));
    assert.ok(!out.includes('No issues.'));
  });
});

// ── CsvPipeline ───────────────────────────────────────────────
describe('CsvPipeline', () => {
  let pipe: CsvPipeline;
  beforeEach(() => { pipe = new CsvPipeline(); });

  it('extract loads 3 records', () => {
    captureOutput(() => (pipe as any).extract());
    assert.strictEqual((pipe as any).records.length, 3);
  });

  it('transform trims whitespace', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    const names = (pipe as any).records.map((r: Row) => r['name']);
    assert.ok(names.includes('Bob'));
    assert.ok(!names.includes('  Bob  '));
  });

  it('transform converts empty string to null', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.strictEqual((pipe as any).records[2]['name'], null);
  });

  it('transform output mentions whitespace trimming', () => {
    captureOutput(() => (pipe as any).extract());
    const lines = captureOutput(() => (pipe as any).transform());
    assert.ok(lines.some(l => l.includes('whitespace trimmed')));
  });
});

// ── JsonPipeline ──────────────────────────────────────────────
describe('JsonPipeline', () => {
  let pipe: JsonPipeline;
  beforeEach(() => { pipe = new JsonPipeline(); });

  it('extract loads 3 records', () => {
    captureOutput(() => (pipe as any).extract());
    assert.strictEqual((pipe as any).records.length, 3);
  });

  it('transform flattens nested objects', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.ok('meta_role' in (pipe as any).records[0]);
    assert.ok(!('meta' in (pipe as any).records[0]));
  });

  it('transform coerces numeric string to number', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.strictEqual((pipe as any).records[0]['age'], 31);
  });
});

// ── XmlPipeline ───────────────────────────────────────────────
describe('XmlPipeline', () => {
  let pipe: XmlPipeline;
  beforeEach(() => { pipe = new XmlPipeline(); });

  it('extract loads 3 records', () => {
    captureOutput(() => (pipe as any).extract());
    assert.strictEqual((pipe as any).records.length, 3);
  });

  it('transform coerces int', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.strictEqual((pipe as any).records[0]['id'], 1);
  });

  it('transform coerces float', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.strictEqual((pipe as any).records[0]['score'], 9.4);
  });

  it('transform coerces bool true', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.strictEqual((pipe as any).records[0]['active'], true);
  });

  it('transform coerces bool false', () => {
    captureOutput(() => (pipe as any).extract());
    captureOutput(() => (pipe as any).transform());
    assert.strictEqual((pipe as any).records[1]['active'], false);
  });
});

// ── run() step order ─────────────────────────────────────────
describe('run() step ordering', () => {
  it('calls steps in correct order', () => {
    const order: string[] = [];
    class SpyPipe extends DataPipeline {
      get name() { return 'Spy'; }
      protected extract()   { order.push('extract');   this.records = [{ id: '1', name: 'X' }]; }
      protected transform() { order.push('transform'); }
      protected override validate() { order.push('validate'); this.errors = []; }
      protected override load()     { order.push('load'); }
      protected override report()   { order.push('report'); }
    }
    captureOutput(() => new SpyPipe().run());
    assert.deepStrictEqual(order, ['extract', 'transform', 'validate', 'load', 'report']);
  });
});

File: patterns/_tests/BehavioralPatterns/TemplateMethod/ruby/unit.test.rb
# ============================================================
#  Unit Tests — Template Method: Data Pipeline (Ruby)
#  Run: rspec unit.test.rb
# ============================================================

require 'rspec'
require 'stringio'
require 'set'
require_relative '../../../../BehavioralPatterns/TemplateMethod/ruby/ruby'

def capture(&block)
  old = $stdout
  $stdout = StringIO.new
  block.call
  $stdout.string.lines.map(&:chomp)
ensure
  $stdout = old
end

# Helper: minimal concrete subclass for testing base behaviour
def make_pipeline(records = [])
  Class.new(DataPipeline) do
    define_method(:name) { "TestPipe" }
    define_method(:extract)   { @records = records }
    define_method(:transform) { }
    define_method(:run_validate) { validate }
    define_method(:run_load)     { load }
    define_method(:run_report)   { report }
    define_method(:errs)         { @errors }
  end.new
end

# ── validate() ───────────────────────────────────────────────
RSpec.describe 'DataPipeline#validate' do
  it 'passes records with id and name' do
    p = make_pipeline([{ id: '1', name: 'Alice' }])
    capture { p.run_validate }
    expect(p.errs).to be_empty
  end

  it 'catches missing name' do
    p = make_pipeline([{ id: '1', name: nil }])
    capture { p.run_validate }
    expect(p.errs).to include(a_string_including('name'))
  end

  it 'catches missing id' do
    p = make_pipeline([{ id: nil, name: 'Alice' }])
    capture { p.run_validate }
    expect(p.errs).to include(a_string_including("'id'"))
  end

  it 'output line contains "records checked"' do
    p = make_pipeline([{ id: '1', name: 'Alice' }])
    lines = capture { p.run_validate }
    expect(lines).to include(a_string_including('records checked'))
  end
end

# ── report() hook ────────────────────────────────────────────
RSpec.describe 'DataPipeline#report hook' do
  it 'prints "No issues." when no errors' do
    p = make_pipeline([])
    out = capture { p.run_report }.join("\n")
    expect(out).to include('No issues.')
  end

  it 'prints warning when errors exist' do
    p = make_pipeline([])
    p.instance_variable_set(:@errors, ["row 0: missing required field 'name'"])
    out = capture { p.run_report }.join("\n")
    expect(out).to include('!')
  end

  it 'can be overridden in a subclass' do
    custom = Class.new(CsvPipeline) do
      def report = puts "CUSTOM REPORT"
    end
    out = capture { custom.new.run }.join("\n")
    expect(out).to include('CUSTOM REPORT')
    expect(out).not_to include('No issues.')
  end
end

# ── CsvPipeline ───────────────────────────────────────────────
RSpec.describe CsvPipeline do
  subject(:pipe) { described_class.new }

  it 'extract loads 3 records' do
    capture { pipe.send(:extract) }
    expect(pipe.instance_variable_get(:@records).size).to eq 3
  end

  it 'transform trims whitespace' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    names = pipe.instance_variable_get(:@records).map { |r| r[:name] }
    expect(names).to include('Bob')
    expect(names).not_to include('  Bob  ')
  end

  it 'transform converts empty string to nil' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    expect(pipe.instance_variable_get(:@records)[2][:name]).to be_nil
  end
end

# ── JsonPipeline ──────────────────────────────────────────────
RSpec.describe JsonPipeline do
  subject(:pipe) { described_class.new }

  it 'extract loads 3 records' do
    capture { pipe.send(:extract) }
    expect(pipe.instance_variable_get(:@records).size).to eq 3
  end

  it 'transform flattens nested hash' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    rec = pipe.instance_variable_get(:@records)[0]
    expect(rec).to have_key(:meta_role)
    expect(rec).not_to have_key(:meta)
  end

  it 'transform coerces numeric string to integer' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    expect(pipe.instance_variable_get(:@records)[0][:age]).to eq 31
  end
end

# ── XmlPipeline ───────────────────────────────────────────────
RSpec.describe XmlPipeline do
  subject(:pipe) { described_class.new }

  it 'extract loads 3 records' do
    capture { pipe.send(:extract) }
    expect(pipe.instance_variable_get(:@records).size).to eq 3
  end

  it 'transform coerces int' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    expect(pipe.instance_variable_get(:@records)[0][:id]).to eq 1
  end

  it 'transform coerces float' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    expect(pipe.instance_variable_get(:@records)[0][:score]).to eq 9.4
  end

  it 'transform coerces bool true' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    expect(pipe.instance_variable_get(:@records)[0][:active]).to be true
  end

  it 'transform coerces bool false' do
    capture { pipe.send(:extract) }
    capture { pipe.send(:transform) }
    expect(pipe.instance_variable_get(:@records)[1][:active]).to be false
  end
end

# ── run() step order ─────────────────────────────────────────
RSpec.describe 'run() step ordering' do
  it 'calls steps in the correct order' do
    order = []
    spy = Class.new(DataPipeline) do
      define_method(:name)      { 'Spy' }
      define_method(:extract)   { order << 'extract'; @records = [{ id: '1', name: 'X' }] }
      define_method(:transform) { order << 'transform' }
      define_method(:validate)  { order << 'validate'; @errors = [] }
      define_method(:load)      { order << 'load' }
      define_method(:report)    { order << 'report' }
    end
    capture { spy.new.run }
    expect(order).to eq %w[extract transform validate load report]
  end
end
